Test suite fails with Python 3.10a3

When switching from Python 3.10a2 -> 3.10a3 the test suite starts failing.


    ======================================================================
    FAIL: test_not_iterable (genshi.template.tests.directives.ForDirectiveTestCase)
    Verify that assignment to nested tuples works correctly.
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "…/genshi/template/tests/directives.py", line 489, in test_not_iterable
        list(tmpl.generate(foo=12))
    TypeError: 'int' object is not iterable

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "…/genshi/template/tests/directives.py", line 505, in test_not_iterable
        self.assertEqual(2, frames[-1].tb_lineno)
    AssertionError: 2 != 61

    ======================================================================
    FAIL: test_error_access_undefined (genshi.template.tests.eval.ExpressionTestCase)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "…/genshi/template/tests/eval.py", line 444, in test_error_access_undefined
        expr.evaluate({})
    genshi.template.eval.UndefinedError: "nothing" not defined

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "…/genshi/template/tests/eval.py", line 458, in test_error_access_undefined
        self.assertEqual(50, frames[-3].tb_lineno)
    AssertionError: 50 != 109

    ======================================================================
    FAIL: test_error_getattr_undefined (genshi.template.tests.eval.ExpressionTestCase)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "…/genshi/template/eval.py", line 301, in lookup_attr
        val = getattr(obj, key)
    AttributeError: 'Something' object has no attribute 'nil'

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "…/genshi/template/eval.py", line 307, in lookup_attr
        val = obj[key]
    TypeError: 'Something' object is not subscriptable

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "…/genshi/template/tests/eval.py", line 467, in test_error_getattr_undefined
        expr.evaluate({'something': Something()})
    genshi.template.eval.UndefinedError: <Something> has no member named "nil"

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "…/genshi/template/tests/eval.py", line 482, in test_error_getattr_undefined
        self.assertEqual(50, frame.tb_lineno)
    AssertionError: 50 != 109

    ======================================================================
    FAIL: test_error_getitem_undefined_string (genshi.template.tests.eval.ExpressionTestCase)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "…/genshi/template/eval.py", line 318, in lookup_item
        return obj[key]
    TypeError: 'Something' object is not subscriptable

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "…/genshi/template/tests/eval.py", line 491, in test_error_getitem_undefined_string
        expr.evaluate({'something': Something()})
    genshi.template.eval.UndefinedError: <Something> has no member named "nil"

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "…/genshi/template/tests/eval.py", line 506, in test_error_getitem_undefined_string
        self.assertEqual(50, frame.tb_lineno)
    AssertionError: 50 != 109

    ----------------------------------------------------------------------
    Ran 886 tests in 0.547s

    FAILED (failures=4)


Initial report: https://bugzilla.redhat.com/show_bug.cgi?id=1907400

Similar failures with Python 3.10rc2 with one minor difference:

    …
      File "…/genshi/template/tests/eval.py", line 506, in test_error_getitem_undefined_string
        self.assertEqual(50, frame.tb_lineno)
    AssertionError: 50 != -1

--- a/genshi/compat.py
+++ b/genshi/compat.py
@@ -99,6 +99,13 @@ else:
 
 
     def build_code_chunk(code, filename, name, lineno):
+        if hasattr(code, 'replace'):
+            # Python 3.8+
+            return code.replace(
+                co_filename=filename,
+                co_name=name,
+                co_firstlineno=lineno,
+            )
         params =  [0, code.co_nlocals, code.co_kwonlyargcount,
                   code.co_stacksize, code.co_flags | 0x0040,
                   code.co_code, code.co_consts, code.co_names,

--- a/genshi/core.py
+++ b/genshi/core.py
@@ -20,7 +20,8 @@ import operator
 
 import six
 
-from genshi.util import plaintext, stripentities, striptags, stringrepr
+from genshi.compat import stringrepr
+from genshi.util import stripentities, striptags
 
 __all__ = ['Stream', 'Markup', 'escape', 'unescape', 'Attrs', 'Namespace',
            'QName']

--- a/genshi/filters/tests/i18n.py
+++ b/genshi/filters/tests/i18n.py
@@ -12,7 +12,6 @@
 # history and logs, available at http://genshi.edgewall.org/log/.
 
 from datetime import datetime
-import doctest
 from gettext import NullTranslations
 import unittest
 

--- a/genshi/filters/tests/test_html.py
+++ b/genshi/filters/tests/test_html.py
@@ -11,7 +11,6 @@
 # individuals. For the exact contribution history, see the revision
 # history and logs, available at http://genshi.edgewall.org/log/.
 
-import doctest
 import unittest
 
 import six

--- a/genshi/filters/tests/transform.py
+++ b/genshi/filters/tests/transform.py
@@ -12,7 +12,6 @@
 # history and logs, available at http://genshi.edgewall.org/log/.
 
 import doctest
-from pprint import pprint
 import unittest
 
 import six

--- a/genshi/input.py
+++ b/genshi/input.py
@@ -22,8 +22,6 @@ from xml.parsers import expat
 import six
 from six.moves import html_entities as entities, html_parser as html
 
-import six
-
 from genshi.core import Attrs, QName, Stream, stripentities
 from genshi.core import START, END, XML_DECL, DOCTYPE, TEXT, START_NS, \
                         END_NS, START_CDATA, END_CDATA, PI, COMMENT

--- a/genshi/output.py
+++ b/genshi/output.py
@@ -20,7 +20,7 @@ import re
 
 import six
 
-from genshi.core import escape, Attrs, Markup, Namespace, QName, StreamEventKind
+from genshi.core import escape, Attrs, Markup, QName, StreamEventKind
 from genshi.core import START, END, TEXT, XML_DECL, DOCTYPE, START_NS, END_NS, \
                         START_CDATA, END_CDATA, PI, COMMENT, XML_NAMESPACE
 

--- a/genshi/template/base.py
+++ b/genshi/template/base.py
@@ -15,7 +15,6 @@
 
 from collections import deque
 import os
-import sys
 
 import six
 

--- a/genshi/template/directives.py
+++ b/genshi/template/directives.py
@@ -19,8 +19,7 @@ from genshi.core import QName, Stream
 from genshi.path import Path
 from genshi.template.base import TemplateRuntimeError, TemplateSyntaxError, \
                                  EXPR, _apply_directives, _eval_expr
-from genshi.template.eval import Expression, ExpressionASTTransformer, \
-                                 _ast, _parse
+from genshi.template.eval import Expression, _ast, _parse
 
 __all__ = ['AttrsDirective', 'ChooseDirective', 'ContentDirective',
            'DefDirective', 'ForDirective', 'IfDirective', 'MatchDirective',

--- a/genshi/template/interpolation.py
+++ b/genshi/template/interpolation.py
@@ -16,7 +16,6 @@ parts that are literal strings, and others that are Python expressions.
 """
 
 from itertools import chain
-import os
 import re
 from tokenize import PseudoToken
 

--- a/genshi/template/markup.py
+++ b/genshi/template/markup.py
@@ -15,7 +15,7 @@
 
 from itertools import chain
 
-from genshi.core import Attrs, Markup, Namespace, Stream, StreamEventKind
+from genshi.core import Attrs, Markup, Namespace, Stream
 from genshi.core import START, END, START_NS, END_NS, TEXT, PI, COMMENT
 from genshi.input import XMLParser
 from genshi.template.base import BadDirectiveError, Template, \

--- a/genshi/template/text.py
+++ b/genshi/template/text.py
@@ -35,7 +35,6 @@ from genshi.template.base import BadDirectiveError, Template, \
                                  TemplateSyntaxError, EXEC, INCLUDE, SUB
 from genshi.template.eval import Suite
 from genshi.template.directives import *
-from genshi.template.directives import Directive
 from genshi.template.interpolation import interpolate
 
 __all__ = ['NewTextTemplate', 'OldTextTemplate', 'TextTemplate']

--- a/genshi/tests/core.py
+++ b/genshi/tests/core.py
@@ -11,13 +11,12 @@
 # individuals. For the exact contribution history, see the revision
 # history and logs, available at http://genshi.edgewall.org/log/.
 
-import doctest
 import pickle
 import unittest
 
 from genshi import core
 from genshi.core import Markup, Attrs, Namespace, QName, escape, unescape
-from genshi.input import XML, ParseError
+from genshi.input import XML
 from genshi.compat import StringIO, BytesIO, IS_PYTHON2
 from genshi.tests.test_utils import doctest_suite
 

--- a/genshi/tests/input.py
+++ b/genshi/tests/input.py
@@ -11,8 +11,6 @@
 # individuals. For the exact contribution history, see the revision
 # history and logs, available at http://genshi.edgewall.org/log/.
 
-import doctest
-import sys
 import unittest
 
 from genshi.core import Attrs, Stream

--- a/genshi/tests/output.py
+++ b/genshi/tests/output.py
@@ -11,9 +11,7 @@
 # individuals. For the exact contribution history, see the revision
 # history and logs, available at http://genshi.edgewall.org/log/.
 
-import doctest
 import unittest
-import sys
 
 from genshi.core import Attrs, Markup, QName, Stream
 from genshi.input import HTML, XML

--- a/genshi/tests/path.py
+++ b/genshi/tests/path.py
@@ -11,7 +11,6 @@
 # individuals. For the exact contribution history, see the revision
 # history and logs, available at http://genshi.edgewall.org/log/.
 
-import doctest
 import unittest
 
 from genshi.core import Attrs, QName

--- a/genshi/tests/util.py
+++ b/genshi/tests/util.py
@@ -11,7 +11,6 @@
 # individuals. For the exact contribution history, see the revision
 # history and logs, available at http://genshi.edgewall.org/log/.
 
-import doctest
 import unittest
 
 from genshi import util

--- a/genshi/util.py
+++ b/genshi/util.py
@@ -19,8 +19,6 @@ from six.moves import html_entities as entities
 
 import six
 
-from .compat import stringrepr
-
 __docformat__ = 'restructuredtext en'
 
 

--- a/genshi/util.py
+++ b/genshi/util.py
@@ -119,7 +119,6 @@ class LRUCache(dict):
 
     def _manage_size(self):
         while len(self._dict) > self.capacity:
-            olditem = self._dict[self.tail.key]
             del self._dict[self.tail.key]
             if self.tail != self.head:
                 self.tail = self.tail.prv

--- a/setup.py
+++ b/setup.py
@@ -12,11 +12,8 @@
 # individuals. For the exact contribution history, see the revision
 # history and logs, available at http://genshi.edgewall.org/log/.
 
-from distutils.cmd import Command
 from distutils.command.build_ext import build_ext
 from distutils.errors import CCompilerError, DistutilsPlatformError
-import doctest
-from glob import glob
 import os
 try:
     from setuptools import setup, Extension

